frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BaseResetBranchToRemoteAction.java

```diff
@@ -0,0 +1,179 @@
+    } else {
+      presentation.setEnabled(false);
+      presentation.setDescription("Reset disabled because ${getActionType()} branch is untracked");
+    }
+  }
+
+  @Override
+  @UIEffect
+  public void actionPerformed(AnActionEvent anActionEvent) {
+    LOG.debug("Performing");
+
+    Project project = getProject(anActionEvent);
+    var gitRepository = getSelectedVcsRepository(anActionEvent);
+    var branchName = getNameOfBranchUnderAction(anActionEvent);
+    var macheteRepository = getGitMacheteRepository(anActionEvent);
+
+    if (branchName.isDefined()) {
+      if (gitRepository.isDefined()) {
+        if (macheteRepository.isDefined()) {
+          doResetToRemoteWithKeep(project, gitRepository.get(), branchName.get(), macheteRepository.get(),
+              anActionEvent);
+        } else {
+          LOG.error("Skipping the action because can't get Git Machete repository");
+          VcsNotifier.getInstance(project).notifyError(VCS_NOTIFIER_TITLE,
+              "Internal error occurred. For more information see IDE log file");
+        }
+      } else {
+        LOG.warn("Skipping the action because no VCS repository is selected");
+        VcsNotifier.getInstance(project).notifyWarning(VCS_NOTIFIER_TITLE,
+            "Skipping the action because no VCS repository is selected");
+      }
+    } else {
+      LOG.warn("Skipping the action because name of branch to reset is undefined");
+      VcsNotifier.getInstance(project).notifyError(VCS_NOTIFIER_TITLE,
+          "Internal error occurred. For more information see IDE log file");
+    }
+  }
+
+  protected void doResetToRemoteWithKeep(Project project, GitRepository gitRepository, String branchName,
+      IGitMacheteRepository macheteRepository, AnActionEvent anActionEvent) {
+
+    new Task.Backgroundable(project, TASK_TITLE, /* canBeCancelled */ true) {
+
+      @Override
+      public void run(ProgressIndicator indicator) {
+        LOG.debug(() -> "Resetting '${branchName}' branch");
+        try (AccessToken ignored = DvcsUtil.workingTreeChangeStarted(project, TASK_TITLE)) {
+          GitLineHandler resetHandler = new GitLineHandler(myProject, gitRepository.getRoot(), GitCommand.RESET);
+          resetHandler.addParameters("--keep");
+
+          Option<BaseGitMacheteBranch> branchOption = macheteRepository.getBranchByName(branchName);
+          assert branchOption.isDefined() : "Can't get branch '${branchName}' from Git Machete repository";
+          Option<IGitMacheteRemoteBranch> remoteTrackingBranchOption = branchOption.get().getRemoteTrackingBranch();
+          if (remoteTrackingBranchOption.isDefined()) {
+            resetHandler.addParameters(remoteTrackingBranchOption.get().getPointedCommit().getHash());
+          } else {
+            String message = "Branch '${branchName}' doesn't have remote tracking branch, so cannot be reset";
+            LOG.warn(message);
+            VcsNotifier.getInstance(project).notifyWarning(VCS_NOTIFIER_TITLE, message);
```

Comment:
Does it throw an error? Shouldn't we have an explicit `return` here?

Reply:
Yeas, again, my fault...